File: Data Structures & Algorithms/string-encode-and-decode/submission-1.cpp
class Solution {
public:

    string encode(vector<string>& strs) {
        string ans;
        for(string s:strs){
            ans+=to_string(s.size())+"#"+s;
        }
        return ans;
    }

    vector<string> decode(string s) {
        int i = 0;
        vector<string> ans;
        while(i<s.size()){
            int j = i;
            while(s[j]!='#')
                j++;
            int len = stoi(s.substr(i,j-i));
            j++;
            ans.push_back(s.substr(j,len));
            i=j+len;
        }
        return ans;
    }
};

File: Data Structures & Algorithms/anagram-groups/submission-1.java
class Solution {

    static String getHash(String s){
        StringBuilder hash = new StringBuilder();
        int[] freq = new int[26];

        for(char c:s.toCharArray()){
            freq[c-'a']++;
        }
        for(int i = 0; i<26; i++){
            hash.append(freq[i]);
            hash.append("$");
        }
        return hash.toString();
    }
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> ans = new ArrayList<>();
        HashMap<String, Integer> map = new HashMap<>();

        for(int i = 0; i<strs.length; i++){
            String key = getHash(strs[i]);

            if(!map.containsKey(key)){
                map.put(key, ans.size());
                ans.add(new ArrayList<>());
            }
            ans.get(map.get(key)).add(strs[i]);
        } 
        return ans;
    }
}
